rag_experiment_accelerator/run/index.py

```diff
@@ -42,68 +112,432 @@ def run(config_dir: str, data_dir: str = "data", filename: str = "config.json")
-                    for ef_search in config.EF_SEARCHES:
-                        index_name = get_index_name(
-                            config.NAME_PREFIX,
-                            chunk_size,
-                            overlap,
-                            embedding_model.name,
-                            ef_construction,
-                            ef_search,
-                        )
-                        data_load = []
-                        for doc in all_docs:
-                            for value in doc.values():
-                                chunk_dict = {
-                                    "content": value,
-                                    "content_vector": embedding_model.generate_embedding(
-                                        chunk=str(pre_process.preprocess(value))
-                                    ),
-                                }
-                                data_load.append(chunk_dict)
-                        upload_data(
-                            chunks=data_load,
-                            service_endpoint=service_endpoint,
-                            index_name=index_name,
-                            search_key=key,
-                            embedding_model=embedding_model,
-                            azure_oai_deployment_name=config.AZURE_OAI_CHAT_DEPLOYMENT_NAME,
-                        )
+
+            futures = {
+                executor.submit(embed_chunk, pre_process, embedding_model, doc): doc
+                for doc in chunks
+            }
+
+            for future in as_completed(futures):
+                doc = futures[future]
+                try:
+                    chunk_dict = future.result()
+                except Exception as exc:
+                    logger.error(
+                        f"embed_chunk generated an exception: {exc} for doc {doc}"
+                    )
+                else:
+                    embedded_chunks.append(chunk_dict)
+                progress_bar.update(1)
+
+    if config.OVERRIDE_CONTENT_WITH_SUMMARY:
+        for chunk in chunks:
+            if "summary" in chunk:
+                chunk["content"] = chunk["summary"]
+                chunk["content_vector"] = chunk["summary_vector"]
+                chunk["summary"] = ""
+                chunk["summary_vector"] = []
+            else:
+                logger.warn("summary was not generated")
+
+    return embedded_chunks
+
+
+def embed_chunk(pre_process, embedding_model, chunk):
+    """
+    Generates an embedding for a chunk of content.
+
+    This function takes a chunk of content, preprocess it and generates an
+    embedding for it using the `generate_embedding` function.
+    The generated embedding is then added to the chunk dictionary under the
+    key "content_vector".
+
+    Args:
+        pre_process (object): An object with a method for preprocessing text.
+                embedding_model (object): The embedding model which was created
```

Comment:
is this line indented too much?

Reply:
The indenting still looks wrong (in the PR review UI):

![image](https://github.com/microsoft/rag-experiment-accelerator/assets/1524451/d3f61196-7862-42ad-875d-ac26d274d89e)

Reply:
fixed